src/controller.ts

```diff
@@ -200,20 +199,23 @@ export class Controller {
       throw new Error("Invalid compilation result");
     }
 
-    let mostDeepSigCell = cell.refs[1];
-
-    while (true) {
-      if (mostDeepSigCell.refs.length === 0) {
-        break;
-      }
-      mostDeepSigCell = mostDeepSigCell.refs[0];
-    }
-
-    const { sigCell } = signatureCell(cell.refs[0], this.keypair);
-    mostDeepSigCell.refs.push(sigCell);
+    const slice = cell.beginParse();
+    const msgToSign = slice.loadRef();
+    const { sigCell } = signatureCell(msgToSign, this.keypair);
+    let child = storeSign(slice.loadRef(), sigCell);
 
     return {
-      msgCell: cell.toBoc(),
+      msgCell: slice.asBuilder().storeRef(msgToSign).storeRef(child).asCell().toBoc(),
     };
+
+    function storeSign(node: Cell, sigCell: Cell): Cell {
+      const slice = node.beginParse();
+      if (slice.remainingRefs > 0) {
+        const child = storeSign(slice.loadRef(), msgToSign);
```

Comment:
1 - is it intentional that we store `msgToSign` and not `sigCell`?

Reply:
about this - the purpose is to store the signature, not the payload so this doesn't make sense to me